lib/workers/branch/index.ts

```diff
@@ -320,6 +326,60 @@ export async function processBranch(
     } else {
       logger.debug('No updated lock files in branch');
     }
+
+    if (global.trustLevel === 'high') {
+      logger.debug('Executing post-change hooks');
+      if (
+        is.nonEmptyArray(config.postUpgradeTasks) &&
+        is.nonEmptyArray(global.allowedPostUpgradeTasks)
+      ) {
+        for (const cmd of config.postUpgradeTasks) {
+          if (
+            !_.some(global.allowedPostUpgradeTasks, (pattern: string) =>
+              cmd.match(pattern)
+            )
+          ) {
+            logger.debug(
+              { cmd, allowedPostUpgradeTasks: global.allowedPostUpgradeTasks },
+              'Post-upgrade task did not match any on allowed list'
+            );
+          } else {
+            logger.debug({ cmd }, 'Executing post-change hook');
+
+            const execResult = await exec(cmd, {
+              cwd: config.localDir,
+            });
+
+            logger.debug(execResult, 'Executed post-change hook');
+          }
+        }
+      }
+
+      if (is.nonEmptyArray(config.postUpgradeFiles)) {
+        const files = find.fileSync(config.localDir);
+
+        for (const filePath of files) {
+          const relativePath = filePath.substring(config.localDir.length + 1);
+
+          if (!relativePath.startsWith('.git/')) {
+            for (const pattern of config.postUpgradeFiles) {
+              if (micromatch.isMatch(relativePath, pattern)) {
+                logger.debug(
+                  { file: filePath, pattern },
+                  'Post upgrade file saved'
+                );
+                const existingContent = await readFile(filePath);
+                config.updatedArtifacts.push({
```

Comment:
Unless I've misunderstood, this will push files regardless of whether they have changed or not. Do you think the logic we have for "vendoring" files (e.g. [in Go Modules](https://github.com/renovatebot/renovate/blob/5b1cd403ab2db660c2c207210efb8a18de218133/lib/manager/gomod/artifacts.ts#L141-L161)) could be used instead? i.e. use git to get a list of changed files then filter them against the list of postUpgradeFiles globs?

Reply:
Not exactly. Renovate will do extra work and store a copy of any matching files in memory, restoring them later. `git add` is then called which would filter out any files which are unchanged. I've followed your suggestion anyway and used the same method as in Go Modules because it will handle deleted files, whereas this does not.